Move ShaderCache shader defs into pipelines

Shfty opened the pull request:
# Objective

- `NO_STORAGE_BUFFERS_SUPPORT`, `NO_ARRAY_TEXTURES_SUPPORT` and `SIXTEEN_BYTE_ALIGNMENT` are injected by `ShaderCache::get`, which occurs after pipeline specialization, and therefore after `Material` gets the chance to inspect or mutate them in `Material::specialize`.
- This prevents SPIR-V usage in `Material`, due to ShaderProcessor asserting that shader_defs is empty.

## Solution

- Moved the aforementioned defs into a public `platform_shader_defs` function in `bevy_render`, which is used by root `SpecializedMeshPipeline` implementors (i.e. ones that do not compose another `SpecializedMeshPipeline`) to initialize their shader defs.
- Fixes #7771
- Alternative solution to #7772 

---

## Changelog
- Root `SpecializedMeshPipeline` implementors have been given a new `shader_refs` member, which is initialized from `RenderDevice` via `platform_shader_defs` in their `FromWorld` implementation.
- In turn, this is used to initialize the local `shader_defs` vec in their `SpecializedMeshPipeline::specialize` implementation before other values are pushed into it.

## Migration Guide

Third-party root `SpecializedMeshPipeline` implementors (i.e. ones that don't rely on `MeshPipeline` or such) will need  to initialize their shader defs using `platform_shader_defs` and a `RenderDevice` reference in order to receive the mentioned defines.

## Notes

This could be refactored to inject shader defs during `SpecializedMeshPipelines::specialize` and `SpecializedComputePipelines::specialize` in order to decouple root `Specialized*Pipeline` implementors from having to the `platform_shader_defs` function.

However, this would involve making the `Specialized*Pipelines` structs authoritative over the baseline set of defs, then passing them into the specialization chain via a new param on the `specialize` trait method. I felt this would be more destructive on account of the API change, so opted not to pursue this implementation.

Playing devil's advocate, an API change is arguably a more explicit way of signaling that previously global behaviour now needs to be implemented locally. Whether `Specialized*Pipeline` is sufficiently front-facing to warrant this I'm not sure, so will defer to review on that front.

superdump reviewed (commented):
One problem I see with this approach is that it could be easy to forget to initialise one’s shader def Vec with these shader defs. It feels perhaps a little easy to do the wrong thing at the moment.

In that sense it might be easier to do the right thing if the shader defs were passed into the specialisation function. What do you think @Shfty @robtfm ?

Shfty commented:
@superdump Yeah, I agree - passing shader defs in from outside would remove the footgun of having the user magic them up from inside their own impl. If API change is acceptable, I'm happy to go through and rework this appropriately.

Also, now #7771 has been merged, should we consider restoring the `ShaderProcessor` def count check as part of this PR? I'm not sure of the concensus on how valuable the previous "must have no defs regardless of format" semantic was, but this would be a good opportunity to bring it back if it's desirable.

Shfty commented:
@superdump @robtfm I've reworked this PR to make `PipelineCache` the single source of truth for these platform-level shader defs, propagating it down the specialization chain via newly-added trait parameters.

Overall, I'm much happier with this approach; it slots comfortably into the existing data structure, centralizes the defs in question, avoids adding unnecessary fields to `Specialized*Pipeline` implementors, and avoids the footgun of requiring consuming code to call an arbitrary method in their construction routine.

As such, this is now an explicitly breaking change, which will better signal the need for migration to existing implementors.

I've updated the top post with the appropriate outline, changelog and migration guide.

robtfm commented:
i haven't been through in detail but i wonder if you have a feeling for how this interacts with what i proposed in #5703 - 
in that pr, we can add shader defs to importable (wgsl or glsl) shader modules, so that when you import a module into your shader you get its defs as well. then (i didn't do this in that pr yet but) things like `NO_STORAGE_ARRAYS` would be associated with a `platform.wgsl` module that any shader that needs it could import.

i suspect this would solve your problem as well since those platform defs then woudn't be present in any spirv shaders since there's no mechanism to use imports with spirv anyway.

Shfty commented:
(Also, I'm realising that much of this context is absent from the top post, having mainly been discussed in Discord. Will edit to amend.)

robtfm commented:
thanks, that context makes it clearer for me. since i'm late i'm going to summarize to make sure we are on the same page

- shader defs consist of platform/device defs (`NO_STORAGE_ARRAYS`) and arbitrary user defs (`TONEMAP_METHOD_XYZ`)
- all these defs are used in the preprocessor within the shader cache to customise non-spirv shaders (to handle availability of features like storage arrays, and to statically select other custom functionality)
- also in non-spirv context, shader defs can be defined in shaders via the '#define' directive, so the full set of shader defs is not knowable until the shader is built by the shader cache (since #imports can be conditional on defs as well), so they can't really be used outside of the shader scope. in the current rust code the only read of the defs is in the preprocessor itself, and (now removed?) for spirv checking that there are no defs

if i understood correctly, what you're suggesting is 
- move the platform/device defs into the pipeline cache so that they can be passed to `Pipeline::specialize`
- access the defs in `Pipeline::specialize` in order to select a spirv entrypoint based on them
- continue to handle non-spirv defs in the shader cache's preprocessor, just sourcing them from the pipeline cache instead

so in the non-spirv case, we would still have a variant being "selected"/built by the shader cache + preprocessor based on what defs are passed in from pipeline cache plus specialize, and in the spirv case we have the variant being selected by the pipeline itself, based on the (platform-subset-of) defs generated by the pipeline cache, and the defs that are output from specialize are ignored.

it feels pretty convoluted to me to have the usage location of the defs vary like this depending on the shader type. in particular i think it would make on-the-fly compiling / caching of shaders (from naga to spirv) harder in the future, since pipelines would need to know ahead of time what "style" of shader they are using, where currently they just do their specializing job regardless of the shader format. it would also maybe mean you can't share spirv shaders so easily between pipelines by just varying the defs that the pipeline outputs.

i think the better answer is to use the defs in the shader cache itself to select the variant somehow for spirv, similarly to what we do for non-spirv. i don't know exactly how to do that - perhaps the Shader::Spirv variant could have a lookup based on entry point plus shader defs that the shader cache uses to select a final entry point, or something like that? or maybe there's a better mechanism to operate on spirv itself like there is with naga, to actually modify the shader code somehow. i'm not sure exactly how it should look but i would need some strong reason why Pipeline::specialize is the right place to do this for spirv, rather than putting that functionality alongside the equivalent non-spirv functionality in the shader cache itself.

Shfty commented:
> thanks, that context makes it clearer for me. since i'm late i'm going to summarize to make sure we are on the same page
> 
> * shader defs consist of platform/device defs (`NO_STORAGE_ARRAYS`) and arbitrary user defs (`TONEMAP_METHOD_XYZ`)
> * all these defs are used in the preprocessor within the shader cache to customise non-spirv shaders (to handle availability of features like storage arrays, and to statically select other custom functionality)
> * also in non-spirv context, shader defs can be defined in shaders via the '#define' directive, so the full set of shader defs is not knowable until the shader is built by the shader cache (since #imports can be conditional on defs as well), so they can't really be used outside of the shader scope. in the current rust code the only read of the defs is in the preprocessor itself, and (now removed?) for spirv checking that there are no defs

Indeed - though I consider definitions made in shader code via `#define` to be outside the scope of this change, since you could do the equivalent in a Rust shader using `#[cfg()]` or such and be similarly unable to share it across the binary / textual boundary.

My "All shader defs" bulletpoint is probably a little broad in that sense!

> if i understood correctly, what you're suggesting is
> 
> * move the platform/device defs into the pipeline cache so that they can be passed to `Pipeline::specialize`
> * access the defs in `Pipeline::specialize` in order to select a spirv entrypoint based on them
> * continue to handle non-spirv defs in the shader cache's preprocessor, just sourcing them from the pipeline cache instead

Somewhat. The defs are accessed in `Material::specialize` in my case, which is downstream of `Specialized*Pipeline::specialize`, but the broader point is to make them generally available (as per pipeline-sourced defs) rather than exclusive to code downstream of `ShaderCache` (i.e. WGSL / GLSL).

I'm not proposing to codify bevy SPIR-V handling as part of this PR. That's beyond scope; the intent here is to open the door for user code to codify its own SPIR-V handling in the absence of official support for something like `rust-gpu` (which is out-of-scope in the near term as per #5634.)

> so in the non-spirv case, we would still have a variant being "selected"/built by the shader cache + preprocessor based on what defs are passed in from pipeline cache plus specialize, and in the spirv case we have the variant being selected by the pipeline itself, based on the (platform-subset-of) defs generated by the pipeline cache, and the defs that are output from specialize are ignored.

As-is, SPIR-V has no intrinsic (or bevy-provided) way to specialize. `naga` translation notwithstanding, what you put in is what you get, with an error + non-rendering material if whatever's compiled in doesn't align with the configured pipeline.

What you describe is effectively the case for SPIR-V in the existing codebase outside of `ShaderCache` late-injecting its definitions; pipeline-sourced defs propagate through the specialization hierarchy, are augmented by `ShaderCache`, and end up getting ignored by `ShaderProcessor` because it can't operate on a binary.

> it feels pretty convoluted to me to have the usage location of the defs vary like this depending on the shader type. in particular i think it would make on-the-fly compiling / caching of shaders (from naga to spirv) harder in the future, since pipelines would need to know ahead of time what "style" of shader they are using, where currently they just do their specializing job regardless of the shader format. it would also maybe mean you can't share spirv shaders so easily between pipelines by just varying the defs that the pipeline outputs.

`Material::specialize` is the only feasible place for user code to step in and assume responsibility for specializing binary shaders as-is, so I consider it reasonable to provide it access to the full range of engine-side defs.

In terms of internal code, it makes sense that a lower-level solution would be preferable. That implies either exposing a lower level trait that can be implemented to custom-specialize at the `ShaderCache` level, or codifying the way bevy-handles SPIR-V.

> i think the better answer is to use the defs in the shader cache itself to select the variant somehow for spirv, similarly to what we do for non-spirv. i don't know exactly how to do that - perhaps the Shader::Spirv variant could have a lookup based on entry point plus shader defs that the shader cache uses to select a final entry point, or something like that? or maybe there's a better mechanism to operate on spirv itself like there is with naga, to actually modify the shader code somehow. i'm not sure exactly how it should look but i would need some strong reason why Pipeline::specialize is the right place to do this for spirv, rather than putting that functionality alongside the equivalent non-spirv functionality in the shader cache itself.

I put together [a shader def -> entrypoint permutation mapping](https://github.com/Bevy-Rust-GPU/bevy-rust-gpu/blob/master/src/entry_point.rs) in the course of building the trait machinery to specify entrypoints in `bevy-rust-gpu`. It works well enough to cover [the parameters for a Rust reimplementation of `bevy_pbr`'s WGSL](https://github.com/Bevy-Rust-GPU/bevy-pbr-rust/blob/master/src/pbr/entry_points.rs), handling enum-like flags, bool / int costs, and potentially generic types in the future.

However, it's arbitrary, as would be any solution that operates on an equally-arbitrary SPIR-V input. My plugin has the luxury of being a third-party integration that can introduce such a restriction with the understanding that it will be handled automatically by the backend, provided that the user follows instructions.

I don't feel the same is true for bevy, since SPIR-V is intrinsically a black box, and the expectation for an engine is to accept that black box without imposing any additional restrictions. If `ShaderCache`-authority over binary specialization is necessary for proper performance, then it should be exposed in such a way that doesn't restrict usage.

On the topic of a `naga_oil` / transpilation / IR approach, I don't think that's viable as things stand; if nothing else, `naga` is currently not comprehensive enough to process arbitrary valid SPIR-V (see #8131, happy to search out specific upstream issues if needed), which would restrict such a solution to its supported subset.

(I'll post some context on how I arrived at the current approach in a separate message re. justifying the `Pipeline::specialize` approach, as this response is running on :sweat_smile:)

Shfty commented:
Some additional context on how I arrived at the `PipelineCache` approach:

This PR originally avoided API changes by introducing a free-floating function to retrieve the defs in question, which would be called at construction time by any `Specialized*Pipeline` implementors. As that would have introduced an easy footgun of failing to call the function, I retooled it to move construction to the topmost point in the specialization hierarchy and pass the result down from there.

During the initial rework, I tried using a `BaseShaderDefs(Vec<ShaderDefVal>)` resource that would be added to the render world at the same time as `PipelineCache`, construct the defs in `FromWorld`, and be available to the various render world systems that call the top-level `specialize` functions, which then waterfall down through `Specialized*Pipeline` and `Material`, ultimately arriving at `ShaderProcessor`. (i.e. the existing implementation, but with top-level `BaseShaderDefs` replacing an extra `ShaderCache` stage between `Material` and `ShaderProcessor`.)

Unfortunately, this didn't work, as one of the sprite pipeline systems started throwing a "does not implement <system prereq traits>" error at its `add_system` call; I'm not certain, but suspect this may have been down to hitting some internal system param count limit, as it was localized to that specific system (which already had a high param count), and no extra type constraints were present.

Looking for alternatives, I noted that `PipelineCache` was present in every system that would consume `BaseShaderDefs`. Given that `ShaderCache` is a member of `PipelineCache`, it seemed reasonable to move def construction up one level in the type hierarchy in order to solve the ordering issue.

robtfm commented:
i'm basically going to repeat what i said before: I don't think `Pipeline::specialize` is the right place to do shader specialization. the pipeline should not care if the shader is spirv or wgsl or naga IR or anything else. this is for clean separation of responsibilities and is to support compiled shader-caching which is something we definitely want in future.

if you can select (either the whole spirv module or the entry point) based on defs in the shader cache in a way that works for you, then maybe it'll work for everyone. if not, we can iterate on it in future.

> SPIR-V is intrinsically a black box

sure, but bevy's `Shader::Spirv` doesn't have to be

> internal system param count limit

if so you can work around this using a tuple argument like `(q_1, q_2): (Query<&One>, Query<&Two>)`. you could also use the platform resource in a `FromWorld` impl to avoid having to pass extra data to `Pipeline::specialize`. i don't think that's the right approach for shader code branching, but it could be useful for texture format selection or similar.

Shfty commented:
> i'm basically going to repeat what i said before: I don't think `Pipeline::specialize` is the right place to do shader specialization. the pipeline should not care if the shader is spirv or wgsl or naga IR or anything else. this is for clean separation of responsibilities and is to support compiled shader-caching which is something we definitely want in future.

That's clear, but separable from both the issue of the `Material` API's access to bevy-side shader defs being incomplete, and my by-necessity implementation doing something suboptimal in userland because the engine only affords me so much control. 

The former remains a demonstrable bevy-specific problem that should be fixed, or documented on the `Material` trait to ensure that users understand they won't be receiving the full set of definitions relating to the `mesh_view` bindings available to their shaders. To that end, #8190 has been opened to track the issue in proper context.

And the latter can be removed from the equation entirely. Specialization best practices would not have entered the discussion if my touchstone example had been something other than a skunkworks `rust-gpu` project that happens to use `Material::specialize` as its entrypoint to the bevy rendering ecosystem.

Ergo, given that I am not asking bevy to adopt such a mechanism, they are not relevant unless upcoming changes to the shader preprocessor are liable to compromise access to shader definitions in such a way that would prevent their usage. In which case, that would be a regression in the existing `Material` API, and very much counter to the discussion with @superdump that motivated this string of issues and pull requests in the first place.

robtfm commented:
> unless upcoming changes to the associated machinery are liable to reduce or remove access to shader definitions in `Material`

this is exactly the case with the `platform.wgsl` i mentioned above (i'm not *just* being difficult :))

superdump commented:
I haven’t had time to catch up on this discussion nor do I right this moment. I just want to say that I haven’t either had the time to look at @robtfm ’s proposed changes and so I hadn’t considered that my proposal might impact them. I’ve chatted with Rob on Discord a bit to align our thinking and it sounds like this platform definitions approach is going to be generally useful. That said, I haven’t quite understood yet what breaks what, though I trust Rob’s judgement to find a solution that fits the use cases and I hope it won’t be too much bother to adjust things to address them all. I’m not the only authority around rendering parts anymore. :) I’ll catch up when I can.

Shfty commented:
> > unless upcoming changes to the associated machinery are liable to reduce or remove access to shader definitions in `Material`
> 
> this is exactly the case with the `platform.wgsl` i mentioned above (i'm not _just_ being difficult :))

In which case, it would be productive to detail those changes in the appropriate issue / PR / RFC so potentially-controversial regressions - and their accompanying resolutions - can be discussed, before using them as justification against affected PRs. 

#5703 only references shader defs in the capacity of adding new pipeline-level ones to describe which bind group engine-sourced data lives in, which doesn't suggest a level of impact comparable to what you've implied here.

It's very well to be told that my existing entrypoints may be taken away because the approach they support doesn't align with bevy's development trajectory - that's an implicit occupational hazard - but I can only contribute effectively and follow said trajectory if the rules of engagement exist on paper.

>> SPIR-V is intrinsically a black box
>
> sure, but bevy's Shader::Spirv doesn't have to be

Is there an RFC or issue open for discussing potential stratification of bevy's SPIR-V handling? This seems like a controversial change from the perspective of using the format for the sake of total control, so it would be good to contribute to discussion on that front.

> I haven’t had time to catch up on this discussion nor do I right this moment. I just want to say that I haven’t either had the time to look at @robtfm ’s proposed changes and so I hadn’t considered that my proposal might impact them. I’ve chatted with Rob on Discord a bit to align our thinking and it sounds like this platform definitions approach is going to be generally useful. That said, I haven’t quite understood yet what breaks what, though I trust Rob’s judgement to find a solution that fits the use cases and I hope it won’t be too much bother to adjust things to address them all. I’m not the only authority around rendering parts anymore. :) I’ll catch up when I can.

Of course, I understand that there's much to do in many spheres outside my localized `rust-gpu` orbit :slightly_smiling_face:

And it's no bother - I want to make sure the correct solution has been found, or at least progressed toward, when the time comes to give it full consideration.

JMS55 commented:
Another use case I've seen asked for: Having a custom material that all it does is use the standard material shaders, but remove the ENVIRONMENT_MAP_LIGHT shader def to disable environment map lighting for a specific mesh.

BenjaminBrienen commented:
@Shfty can you rebase this?
